cer-graphql/executeQuery.ts

```diff
@@ -0,0 +1,296 @@
+/**
+ * executeQuery.ts
+ * A custom execution function for GraphQL.js, containing validations for
+ * unauthenticated queries.
+*/
+
+import { Request } from "express";
+import { 
+    execute,
+    ExecutionArgs,
+    print,
+    ASTNode,
+    FieldNode,
+    FragmentDefinitionNode,
+    GraphQLSchema,
+    TypeInfo,
+    visit,
+    visitWithTypeInfo
+} from "graphql";
+import { AuthenticationError } from "apollo-server-errors";
+
+const GRAPHQL_INTROSPECTION_FIELDS = [
+    '__Schema',
+    '__Type',
+    '__TypeKind',
+    '__typename',
+    '__Field',
+    '__InputValue',
+    '__EnumValue',
+    '__Directive',
+    'sys',
+    'id',
+  ];
+  
+  // Check whether the user has requested only public fields
+  const ALWAYS_PUBLIC_FIELDS = new Set([
+    'title',
+    'maoriProverb',
+    'summary',
+    'name',
+    'ssoProtected',
+    'searchable',
+    'linkedFrom',
+    'slug',
+    'banner',
+    'icon',
+    ...GRAPHQL_INTROSPECTION_FIELDS
+  ]);
+  
+
+function isProtectedField(fieldName: string, typeName: string) {
+    // A protected field is a field that:
+    // 1. isn't in the ALWAYS_PUBLIC_FIELDS set,
+    // 2. isn't a Contentful "items" field for collection items,
+    // 3. isn't a collection field itself.
+    return !ALWAYS_PUBLIC_FIELDS.has(fieldName) &&
+        !(typeName.endsWith("Collection") && fieldName === "items") &&
+        !fieldName.endsWith("Collection")
+}
+
+type FragmentFieldDepthInfo = {
+    depth: number,
+    name: string
+  };
+
+/**
+ * Given a query document, check that it doesn't ask for protected, nested fields on protected types.
+ * @param document The query document
+ * @param maxDepth number of nested selectionsets a protected field can be in.
+ * @throws AuthenticationError if there are any protected nested fields.
+ */
+function assertNoDeepProtectedFields(document: ASTNode, schema: GraphQLSchema, protectedTypes: Set<string>, maxDepth = 3) {
```

Comment:
This is pretty complicated and I don't know if I understood it all, but I what I did understand seems to make sense.. Could be a good topic for a RAD meeting :)